### PR TITLE
8343153: compiler/codecache/CheckLargePages.java fails on linux with huge pages configured but its number set to 0

### PR DESCRIPTION
**Pull request by dafedafe**
# Issue

The third test of `compiler/codecache/CheckLargePages.java` checks that non-segmented 1GB code cache can be allocated with 1GB large pages.

On linux (the only supported platform) in order to allocate them, 1GB huge pages have to be enabled (checkable from `/proc/meminfo` and `/sys/kernel/mm/hugepages/hugepages-xxxx`) and their number has to be set to >0 (checkable from `/sys/kernel/mm/hugepages/hugepages-xxxx/nr_hugepages`).

If 1GB huge pages are enabled but their number is 0, the test fails because it looks for a string that matches `CodeCache: min=1[gG] max=1[gG] base=[^ ]+ size=1[gG] page_size=1[gG]` but the actual output is `CodeCache: min=1G max=1G base=0x00007f4040000000 size=1G page_size=2M`. This happens because the VM tries to allocate 1GB huge pages but it fails beause the number of allocatable ones is 0 and the VM reverts to smaller large page sizes (2MB).

# Solution

The problem might be attributed to the VM only checking for 1GB huge pages to be supported, not how many there currently are. Nevertheless, this seems to be the correct behaviour, not least because their number can be changed dynamically.
So, the correct thing to do seems to be to "relax" the check made by the test to include both cases:
* when 1GB huge pages are supported and can be allocated correctly
* when 1GB huge pages are supported but cannot be allocated correctly (because there are none available) and the VM reverts to 2MB huge pages (if there are no 2MB pages available the test doesn't run at all).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343153](https://bugs.openjdk.org/browse/JDK-8343153): compiler/codecache/CheckLargePages.java fails on linux with huge pages configured but its number set to 0 (**Bug** - P4)

### Reviewers
 * [Evgeny Astigeevich](https://openjdk.org/census#eastigeevich) (@eastig - Committer)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21757/head:pull/21757` \
`$ git checkout pull/21757`

Update a local copy of the PR: \
`$ git checkout pull/21757` \
`$ git pull https://git.openjdk.org/jdk.git pull/21757/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21757`

View PR using the GUI difftool: \
`$ git pr show -t 21757`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21757.diff">https://git.openjdk.org/jdk/pull/21757.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21757#issuecomment-2447734537)
</details>

### CODE REVIEW

**Comment by dafedafe** [comment]
@eastig I noticed that you are the author of the original `testNonSegmented1GbCodeCacheWith1GbLargePages` test. Could I ask you to have a look at this change? Thanks a lot!

File: test/hotspot/jtreg/compiler/codecache/CheckLargePages.java [hunk 1]
```diff
@@ -111,8 +111,20 @@ private static void testNonSegmented1GbCodeCacheWith1GbLargePages() throws Excep
...
         out.shouldNotContain("CodeHeap 'non-profiled nmethods'");
         out.shouldContain("UseLargePages=1, UseTransparentHugePages=0");
-        out.shouldMatch("CodeCache:  min=1[gG] max=1[gG] base=[^ ]+ size=1[gG] page_size=1[gG]");
+        out.shouldMatch("Large page support enabled\\. Usable page sizes: .*1[gG].*\\. Default large page size: .*\\.");
+        out.shouldMatch(
+                // 1GB large pages configured and available
+                "CodeCache:\\s+min=1[gG] max=1[gG] base=[^ ]+ size=1[gG] page_size=1[gG]|" +
+                // 1GB large pages configured but none available
+                "Failed to reserve and commit memory with given page size\\. " +
+                "req_addr: [^ ]+ size: 1[gG], page size: 1[gG], \\(errno = 12\\)");
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/codecache/CheckLargePages.java:120]: Took me a while to figure that these are `OR` matches due to the `|` hiding at the end of the first line. Would it make sense to update the comment to something like this?
```
                // 1GB large pages configured and available
                "CodeCache:\\s+min=1[gG] max=1[gG] base=[^ ]+ size=1[gG] page_size=1[gG]|" +
                // or 1GB large pages configured but none available
```

Comment (TobiHartmann) [reply test/hotspot/jtreg/compiler/codecache/CheckLargePages.java:120]: Also, isn't there a `CodeCache:\\` line in the output in the failing case as well that should be added here in the OR part?

**Comment by eastig** [comment]
> The problem might be attributed to the VM only checking for 1GB huge pages to be supported, not how many there currently are.

https://bugs.openjdk.org/browse/JDK-8321526

**Comment by eastig** [comment]
> @eastig I noticed that you are the author of the original `testNonSegmented1GbCodeCacheWith1GbLargePages` test. Could I ask you to have a look at this change? Thanks a lot!

`testDefaultCodeCacheWith1GbLargePages` and `testNonSegmented1GbCodeCacheWith1GbLargePages` should only be run if a system provides 1Gb pages. This is mentioned in their names: `...With1GbLargePages`. If there are no 1Gb pages available, the test should not be run.

I suggest to check `/sys/kernel/mm/hugepages/hugepages-1048576kB/nr_hugepages >= 1`. If not, output "Skipping testDefaultCodeCacheWith1GbLargePages and testDefaultCodeCacheWith1GbLargePages, no 1Gb pages available"

**Comment by dafedafe** [comment]
Thanks for looking into it @eastig.

I've changed the test to check for the content of `/sys/kernel/mm/hugepages/hugepages-1048576kB/nr_hugepages >= 1` as you suggested.

File: test/hotspot/jtreg/compiler/codecache/CheckLargePages.java (hunk 1 above)
Comment (dafedafe) [reply test/hotspot/jtreg/compiler/codecache/CheckLargePages.java]: Thanks @TobiHartmann for looking at it.
I've actually changed the test to follow @eastig's suggestion below and reverted these lines to their original state.

**Review by eastig — APPROVED** [review]
Looks good to me.

**Comment by dafedafe** [comment]
Thank you for your reviews @eastig @TobiHartmann.

**Pull request closed without merging**